packages/react-native-codegen/src/parsers/typescript/parseTopLevelType.js

```diff
@@ -0,0 +1,177 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow strict-local
+ * @format
+ */
+
+'use strict';
+
+import type {TypeDeclarationMap} from './utils.js';
+
+export type LegalDefaultValues = string | number | boolean | null;
+
+type TopLevelTypeInternal = {
+  unions: Array<$FlowFixMe>,
+  optional: boolean,
+  defaultValue?: LegalDefaultValues,
+};
+
+export type TopLevelType = {
+  type: $FlowFixMe,
+  optional: boolean,
+  defaultValue?: LegalDefaultValues,
+};
+
+function getValueFromTypes(
+  value: $FlowFixMe,
+  types: TypeDeclarationMap,
+): $FlowFixMe {
+  switch (value.type) {
+    case 'TSTypeReference':
+      if (types[value.typeName.name]) {
+        return getValueFromTypes(types[value.typeName.name], types);
+      } else {
+        return value;
+      }
+    case 'TSTypeAliasDeclaration':
+      return getValueFromTypes(value.typeAnnotation, types);
+    default:
+      return value;
+  }
+}
+
+function isNull(t: $FlowFixMe) {
+  return (
+    t.type === 'TSNullKeyword' ||
+    t.type === 'TSUndefinedKeyword' ||
+    t.type === 'TSVoidKeyword'
+  );
+}
+
+function handleUnionAndParen(
+  type: $FlowFixMe,
+  result: TopLevelTypeInternal,
+  knownTypes?: TypeDeclarationMap,
+): void {
+  switch (type.type) {
+    case 'TSParenthesizedType': {
+      handleUnionAndParen(type.typeAnnotation, result, knownTypes);
+      break;
+    }
+    case 'TSUnionType': {
+      // the order is important
+      // result.optional must be set first
+      for (const t of type.types) {
+        if (isNull(t)) {
+          result.optional = true;
+        }
+      }
+      for (const t of type.types) {
+        if (!isNull(t)) {
+          handleUnionAndParen(t, result, knownTypes);
+        }
+      }
```

Comment:
I would extract this in its own `handleUnionType` function

Reply:
Since it only doing one step of recursion so I think it is fine putting here. But I've extracted the `evaluateLiteral` function.